messagequeue/enqueuer.go

```diff
@@ -0,0 +1,158 @@
+package messagequeue
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/beaconsoftwarellc/gadget/v2/errors"
+)
+
+const (
+	statusStopped  uint32 = 0
+	statusRunning  uint32 = 1
+	statusDraining uint32 = 2
+)
+
+// HandleFailedEnqueue is called when a message fails to enqueue
+type HandleFailedEnqueue func(Enqueuer, *EnqueueMessageResult)
+
+// Enqueuer for inserting tasks into a MessageQueue
+type Enqueuer interface {
+	// Start validates configurations and creates the resources necessary to
+	// handle enqueuing messages.
+	Start(messageQueue MessageQueue) error
+	// Enqueue the passed message. This operation may be
+	// buffered and errors enqueueing will not be available for immediate handling.
+	Enqueue(messages *Message) error
+	// Stop all active workers, drain queues, and free resources.
+	Stop() error
+}
+
+// New enqueuer for inserting messages into a MessageQueue
+func New(options *EnqueuerOptions) Enqueuer {
+	if nil == options {
+		options = NewEnqueuerOptions()
+	}
+	return &enqueuer{
+		options: options,
+		chunker: NewChunker[*Message](options.ChunkerOptions),
+	}
+}
+
+type enqueuer struct {
+	status       atomic.Uint32
+	messageQueue MessageQueue
+	options      *EnqueuerOptions
+	// buffer used by the chunker
+	buffer chan *Message
+	// buffer where we send messages that fail to enqueue that will
+	// be fed to the failedHandler
+	failed chan *EnqueueMessageResult
+	// chunker has the logic for batching our enqueue calls
+	chunker Chunker[*Message]
+	mux     sync.Mutex
+}
+
+func (qr *enqueuer) Start(messageQueue MessageQueue) error {
+	if nil == messageQueue {
+		return errors.New("messageQueue cannot be nil")
+	}
+	qr.mux.Lock()
+	var err error
+	if !qr.status.CompareAndSwap(statusStopped, statusRunning) {
+		qr.mux.Unlock()
+		return errors.New("Enqueuer.Start called while not in state 'Stopped'")
+	}
+	if err = qr.options.Validate(); nil != err {
+		qr.status.Store(statusStopped)
+		qr.mux.Unlock()
+		return err
+	}
+	qr.messageQueue = messageQueue
+	qr.buffer = make(chan *Message)
+	err = qr.chunker.Start(qr.buffer, qr.sendBatch)
+
+	qr.failed = make(chan *EnqueueMessageResult, int(qr.options.FailedBufferSize))
+	go qr.handleFailed()
+	qr.mux.Unlock()
+	if nil != err {
+		qr.Stop()
+	}
+	// set the failure handler to just log if one is not defined
+	if nil == qr.options.FailureHandler {
+		qr.options.FailureHandler = qr.logFailed
+	}
+	return err
+}
+
+func (qr *enqueuer) Stop() error {
+	qr.mux.Lock()
+	if !qr.status.CompareAndSwap(statusRunning, statusStopped) {
+		qr.mux.Unlock()
+		return errors.New("Enqueuer.Stop called while not in state 'Running'")
+	}
+	defer qr.mux.Unlock()
```

Comment:
why not do this prior to the the CompareAndSwap instead of the explicit unlock in the error case?

Reply:
because I refactored and didn't notice, good catch :+1: